File: 2020/day1.py
def part1():
    encountered = set()
    with open("./day1.input") as file_input:
        for line in file_input.readlines():
            num = int(line)
            counterpart = 2020 - num
            if counterpart in encountered:
                print(num * counterpart)
                break
            else:
                encountered.add(num)


def part2():
    numbers = []
    with open("./day1.input") as file_input:
        for line in file_input.readlines():
            num = int(line)
            numbers.append(num)
    n = len(numbers)
    numbers = sorted(numbers)
    lookup_set = set(numbers)
    minimum = numbers[0]
    for i in range(0, n-2):
        x = numbers[i]
        for j in range(i+1, n-1):
            y = numbers[j]
            if x + y + minimum > 2020:
                break
            z = 2020 - x - y
            if z in lookup_set:
                print(x * y * z)
                return


def main():
    part2()


if __name__ == "__main__":
    main()

File: 2024/day1.go
package main

import (
	"bufio"
	"cmp"
	"fmt"
	"log"
	"math"
	"os"
	"slices"
)

func main() {
	file, err := os.Open("day1.input")
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	var xs, ys []int
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		var x, y int
		fmt.Sscanf(line, "%d %d", &x, &y)
		xs = append(xs, x)
		ys = append(ys, y)
	}

	slices.SortFunc(xs, func(a int, b int) int {
		return cmp.Compare(a, b)
	})
	slices.SortFunc(ys, func(a int, b int) int {
		return cmp.Compare(a, b)
	})

	totalDistance := 0
	for i := range xs {
		totalDistance += int(math.Abs(float64(xs[i] - ys[i])))
	}
	fmt.Printf("part1: %d\n", totalDistance)

	counts := make(map[int]int)
	for _, v := range ys {
		counts[v] += 1
	}

	similarityScore := 0
	for _, v := range xs {
		similarityScore += v * counts[v]
	}
	fmt.Printf("part2: %d\n", similarityScore)
}
